### ISSUE TITLE
sql log took long time, how can i make it faster?

### ISSUE THREAD

**Issue by mango-lzp** [dolthub/dolt#4687]
![image](https://user-images.githubusercontent.com/41472923/199467635-bfc9d0dd-4a4f-4c4a-acdc-acc6c94586e9.png)

sql is :  "  select count(*) as \`total\` from \`dolt_log\` inner join \`dolt_diff_i18n\` on \`dolt_log\`.\`commit_hash\` = \`dolt_diff_i18n\`.\`to_commit\` where \`to_originalKey\` = 'asdasd121'; "

![image](https://user-images.githubusercontent.com/41472923/199468862-eba87091-9888-4f05-85ee-4f93a2a75461.png)

i try to create index for join table but i failed.

![image](https://user-images.githubusercontent.com/41472923/199467505-c5fc6bc3-fadd-4dd2-bf68-efef78549c50.png)

**Comment by timsehn** [comment]
Interesting bug. I think this might actually be the `dolt_diff` table. I'll get one of the experts to look at it and make a recommendation.

**Comment by max-hoffman** [comment]
Thanks for the report @mango-lzp. We have a few ideas of what could be happening here, but if you share a couple more details we will be able to repro your specific use case:
1) `dolt version` output
2) `dolt sql -q "select count(*) from dolt_log"` output
3) `dolt sql -q "select count(*) from dolt_diff_i18n"` output

Since the count output is zero, I think we should be able to repro with the 1) appropriate dolt version, and the number of 2) commit rows and 3) diff rows you are trying to join.

If you are on an older dolt version you might see a perf improvement from a HashJoin implementation in one of the more recent releases. We should also automatically be indexing the lookup (dolt_diff tables are read only but we can implicitly index them), but we are not currently. We will follow up with action items soon.

**Comment by mango-lzp** [comment]
![image](https://user-images.githubusercontent.com/41472923/199867441-7f07680c-a908-453f-a836-3d10e21c4486.png)

Thank you for taking time to review my question!

**Comment by timsehn** [comment]
Thanks. @max-hoffman is writing a blog today but he will look at this soon.

**Comment by max-hoffman** [comment]
I will start generating a repo with the same number of commits and diffs for testing. We shipped a join rewrite in Dolt [0.50.11](https://github.com/dolthub/dolt/releases/tag/v0.50.11), which I will compare to `0.50.4`. Might be worth trying on your end also.

**Comment by max-hoffman** [comment]
I made a repo with `6000` rows and `91000` diffs and I found something interesting.

In the dolt version you were using, we neither indexed the diff table nor used hash join to speed up queries. In the new dolt we do use hash joins, but they fail miserably in this case. We try to put 0 rows into the hash map, because no rows match your `'to_originalKey = 'asdasd121'` filter. But then on the next pass fail to notice that we already attempted to fill the cache; we try to repopulate it again because the row count is 0. So we are turning an `O(6000+90000)` query into `O(6000*90000)`.

This will be easy to fix and should make this query ~6000 times faster. I will get something to you by Monday. Long-term we would also like to index history tables to make these faster.

**Comment by max-hoffman** [comment]
I get a 4 second runtime with the same 6000 commit/91000 diff database with this fix: https://github.com/dolthub/go-mysql-server/pull/1382

**Comment by mango-lzp** [comment]
Thanks a lot, I can't wait to upgrade.

**Comment by max-hoffman** [comment]
This release should contain the update. If I missed something or you find more bugs please let me know! We will follow up on diff table indexing in the next couple of months.

**Comment by mango-lzp** [comment]
![}TR5MP (HT42ZE P@2ALV`P](https://user-images.githubusercontent.com/41472923/201625717-d5d9b835-acaa-44f7-9cea-f8d0ef6b4109.png)

I try to upgrade dolt version, but it seems doesn't work.

**Comment by max-hoffman** [comment]
Hi @mango-lzp , sorry to hear you are still seeing issues.

Two things stick out to me:

1) I overlooked that you are on `__LD_1__`, which is going to be broadly 4-5x slower than the `__DOLT__` storage format. We have some [blogs](https://www.dolthub.com/blog/2022-11-01-dolthub-migrate-button/) detailing how to perform the update.

2) This query is a bit different from the last, so I'd expect it to have a slightly different performance profile.

Other customers who migrated to `__DOLT__` (1) have experienced notable system table speedups. I would start there, I think you will see performance similar to what I reported for your first query. I still have my eye on using a lookup index to take this join from `O(m+n)` to `O(min(n,m))`
